Views: Support contributing welcome-views to empty tree view

EstherPerelman opened the pull request:
Signed-off-by: Esther Perelman <esther.perelman@sap.com>

Welcome-views can be contributed by extensions in order to be applied to empty tree views, A view is considered empty if the tree has no children.
see vscode doc: [vs-code-contributes.viewsWelcome](https://code.visualstudio.com/api/references/contribution-points#contributes.viewsWelcome)
#### What it does
Partially solves #7178  #8667  (For now supports contributing to custom-views and 'explorer' view but not the other built-in views: 'scm', 'debugger', 'test').
#### How to test
1. Deploy test extension [run-configuration.zip](https://github.com/eclipse-theia/theia/files/5452263/run-configuration.zip)
When opening it from the left menu - you should see the contributed welcome-view:
<img width="781" alt="Screenshot 2020-10-28 155722" src="https://user-images.githubusercontent.com/71729183/97446028-9691d480-1936-11eb-9e14-96a3148bf2bf.png">
2. You can now contribute more welcome-views to the run-configuration view 
by adding a viewWelcome to any extension package.json under "contributes" key, For example:

   "viewsWelcome": [{
       "view": runConfigurations,
       "contents": "In order to use git features, you can open a folder containing a git repository or clone from a URL.\n[Open Folder](command:vscode.openFolder)\n[Clone Repository](command:git.clone)\nTo learn more about how to use git and source control in VS Code [read our docs](https://aka.ms/vscode-scm).",
      }]

Now it should look like this:
<img width="712" alt="Screenshot 2020-10-28 161547" src="https://user-images.githubusercontent.com/71729183/97448188-0c973b00-1939-11eb-82d8-fc4bb0049db9.png">

You can play around... (add when clause etc...)

3. open theia without opening a workspace/folder - you should see a new welcome view in the Explorer tab, You can contribute additional views-welcome to the Explorer view by specifying "view": "explorer" .

The contentns are rendered like this: 
'/n' - for new line.
'[here goes the label](here goes a command or a link)' - If there's only a command/link in a line without any other text - it will be displayed as a button, Otherwise as a link.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto commented on packages/plugin-ext/src/main/browser/style/tree.css:
```diff
@@ -44,3 +44,28 @@
 }
+
+.theia-tree-view .theia-TreeContainer .theia-WelcomeViewInfo {
+    padding: 7px 20px 0 20px;
```
We should not use hardcoded constants but rather use variables.
The same applies to other css updates included in the pull-request.

vince-fugnitto commented on packages/plugin-ext/src/main/browser/style/tree.css:
```diff
@@ -44,3 +44,28 @@
+
+.theia-tree-view .theia-TreeContainer .theia-WelcomeViewInfo .theia-WelcomeViewButton {
+    width: 100%;
+	max-width: 260px;
+	margin-left: auto;
+	margin-right: auto;
```
Please fix the formatting inconsistencies.

vince-fugnitto commented on packages/plugin-ext/src/main/browser/view/plugin-view-welcome.tsx:
```diff
@@ -0,0 +1,213 @@
+/********************************************************************************
+ * Copyright (C) 2018-2019 Red Hat, Inc. and others.
```
2020 SAP?

vince-fugnitto reviewed (commented):
We can perform a comprehensive review after the next release (tomorrow) :+1:

kittaakos commented on packages/core/src/browser/tree/tree-view-welcome-widget.tsx:
```diff
@@ -0,0 +1,245 @@
+
+    protected renderViewWelcome(): React.ReactNode {
+        return (
+            <div className="theia-WelcomeView">
```
`<div className="theia-WelcomeView">` -> `<div className='theia-WelcomeView'>`

kittaakos replied:
Please change other double quotes to single: `"` -> `'`

kittaakos commented on packages/core/src/browser/tree/tree-view-welcome-widget.tsx:
```diff
@@ -0,0 +1,245 @@
+            const command = href.replace('command:', '');
+            this.commands.executeCommand(command);
+        } else {
+            window.open(href, '_blank');
```
This must go through the `WindowService`. Otherwise, it can be a security issue in Electron.

kittaakos commented on packages/core/src/browser/tree/tree-view-welcome-widget.tsx:
```diff
@@ -0,0 +1,245 @@
+    protected parseLinkedText(text: string): LinkedTextItem[] {
+        const result: LinkedTextItem[] = [];
+
+        const linkRegex = /\[([^\]]+)\]\(((?:https?:\/\/|command:)[^\)\s]+)(?: ("|')([^\3]+)(\3))?\)/gi;
```
It is very important to open a CQ if you have copied code from an open-source project.

https://github.com/microsoft/vscode/blob/573e5145ae3b50523925a6f6315d373e649d1b06/src/vs/base/common/linkedText.ts#L26

kittaakos commented on packages/core/src/browser/tree/tree-view-welcome-widget.tsx:
```diff
@@ -0,0 +1,245 @@
+    visible: boolean;
+}
+
+interface ILink {
+    readonly label: string;
+    readonly href: string;
+    readonly title?: string;
+}
```
This PR requires a CQ for sure:

https://github.com/microsoft/vscode/blob/573e5145ae3b50523925a6f6315d373e649d1b06/src/vs/base/common/linkedText.ts#L8-L12


@marcdumais-work, @amiramw can one of you please help @EstherPerelman with the CQ submission? The changeset works pretty great. Theia needs this addition but we have to clearly document, the idea was taken from VS Code. Thank you!

amiramw replied:
@kittaakos i have never dealt with CQ. is the process documented somewhere?

kittaakos replied:
Yes, it is documented on the Wiki. [Here](https://github.com/eclipse-theia/theia/wiki/Registering-CQs#codecontent-is-copied-from-the-main-vs-code-repo) is the relevant section when the code was copied from VS Code.

marcdumais-work replied:
@amiramw have you opened the CQ? With the increased contributions, it's good that one SAP committer joins the ranks of our "IP champions". Let me know if I have questions. 

If by chance you can find the code you want, already covered by a previous, approved CQ, then we would not need a new one. For example, we often request to copy from `vscode.d.ts`.

amiramw replied:
@marcdumais as i understand since this copied code is from newer version then i need to open a CQ based on https://github.com/eclipse-theia/theia/wiki/Registering-CQs#codecontent-is-copied-from-other-projectsrepos, right?

Where can I see all previous CQs?

amiramw replied:
@kittaakos maybe you can answer?

kittaakos replied:
> Where can I see all previous CQs?

No idea. I bet filtering by label does not show all: https://github.com/eclipse-theia/theia/pulls?q=+is%3Apr+label%3A%22CQ+Required%22+

marcdumais-work replied:
> @marcdumais as i understand since this copied code is from newer version then i need to open a CQ based on https://github.com/eclipse-theia/theia/wiki/Registering-CQs#codecontent-is-copied-from-other-projectsrepos, right?
> 

Yes, I think so. The exception being if the code has been there for a while, in a form you can use, then you are allowed to pick the code from whatever version you want.

> Where can I see all previous CQs?

I am no expert, but I start [here](https://dev.eclipse.org/ipzilla/query.cgi) and usually only do a simple keywork search, e.g. for [vscode](https://dev.eclipse.org/ipzilla/buglist.cgi?query_format=advanced&short_desc_type=allwordssubstr&short_desc=vscode&long_desc_type=substring&long_desc=&bug_file_loc_type=allwordssubstr&bug_file_loc=&keywords_type=allwords&keywords=&emailassigned_to1=1&emailtype1=substring&email1=&emailassigned_to2=1&emailreporter2=1&emailcc2=1&emailtype2=substring&email2=&bugidtype=include&bug_id=&chfieldfrom=&chfieldto=Now&chfieldvalue=&cmdtype=doit&order=Reuse+same+sort+as+last+time&field0-0-0=noop&type0-0-0=noop&value0-0-0=) or whatever I am looking for.

kittaakos replied:
Has anyone already registered the CQ? I do not want to hold back this great change. So if this PR is blocked only by me, let me know.

amiramw replied:
@kittaakos @marcdumais-work I registered CQ https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22793 now. It's my first time so I hope someone can take a look at it.

kittaakos replied:
Great! Thank you so much.

I looked into the attachment, and it looks good to me. I would have created the same archive.🤞

amiramw replied:
@kittaakos the CQ is approved now

kittaakos commented on packages/core/src/browser/tree/tree-view-welcome-widget.tsx:
```diff
@@ -0,0 +1,245 @@
+    @inject(ContextKeyService)
+    protected readonly contextService: ContextKeyService;
+
+    private viewWelcomeNodes: React.ReactNode[] = [];
+    private defaultItem: IItem | undefined;
+    private items: IItem[] = [];
```
Generally, it's better to have `protected` everywhere instead of `private`. Just think about it, what if you would like to quickly customize something on your side, `protected` visibility could come in handy. Please change. Thanks!

kittaakos commented on packages/filesystem/src/browser/file-tree/file-tree-widget.tsx:
```diff
@@ -33,7 +33,7 @@ export const DIR_NODE_CLASS = 'theia-DirNode';
 
 @injectable()
-export class FileTreeWidget extends TreeWidget {
+export class FileTreeWidget extends TreeViewWelcomeWidget {
```
Can we enable the welcome-support in all other trees as well? It isn't FS specific. What do you think, @akosyakov?

kittaakos reviewed (commented):
I tried it with a VS Code extension, and it works perfectly. I look forward to seeing this feature in Theia. Nice job, @EstherPerelman  👍

EstherPerelman commented:
@kittaakos @vince-fugnitto Can you continue reviewing the code?

DukeNgn commented on packages/search-in-workspace/package.json:
```diff
@@ -9,7 +9,6 @@
     "@theia/navigator": "^1.7.0",
     "@theia/process": "^1.7.0",
     "@theia/workspace": "^1.7.0",
-    "minimatch": "^3.0.4",
```
I believe you made this change by mistake when rebasing. Please revert it as `minimatch` is required for siw.

amiramw commented:
@kittaakos @vince-fugnitto other then the CQ do you have more comments?

vince-fugnitto reviewed (commented):
The changes look good to me, I verified with the simple extension but not the additional configuration as I do not have the source code.

I'm wondering about the following https://github.com/eclipse-theia/theia/pull/8678/files#r515101313 however. I'll add reviewers who are more familiar with the plugin-system to see if they have any comments as well.

kittaakos reviewed (commented):
It works nicely, very good job. I added a few comments.

One question; if _Clone Repository_ is not supported when there is no opened workspace, why do we show it?

![Screen Shot 2020-11-20 at 10 08 08](https://user-images.githubusercontent.com/1405703/99781907-c0df4800-2b18-11eb-8fbe-d6f59be66ab1.jpg)

EstherPerelman commented:
> It works nicely, very good job. I added a few comments.
> 
> One question; if _Clone Repository_ is not supported when there is no opened workspace, why do we show it?
> 
> ![Screen Shot 2020-11-20 at 10 08 08](https://user-images.githubusercontent.com/1405703/99781907-c0df4800-2b18-11eb-8fbe-d6f59be66ab1.jpg)

@amiramw What do you think?

EstherPerelman commented on packages/plugin-ext/src/hosted/node/scanners/scanner-theia.ts:
```diff
@@ -483,6 +495,35 @@ export class TheiaPluginScanner implements PluginScanner {
+        const pluginViewsIds: string[] = [];
+        if (views) {
+            for (const location in views) {
+                if (Object.prototype.hasOwnProperty.call(views, location)) {
+                    const viewsIds = views[location].map(view => view.id);
+                    pluginViewsIds.push(...viewsIds);
+                }
+            };
```
I just got Eslint error:
The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype
[.eslintguard-for-in](https://eslint.org/docs/rules/guard-for-in)
Should I disable it?

EstherPerelman commented on packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:
```diff
@@ -132,6 +134,16 @@ export class PluginViewRegistry implements FrontendApplicationContribution {
+                this.registerViewWelcome({
+                    view: 'explorer',
+                    // eslint-disable-next-line max-len
+                    content: 'You have not yet opened a folder.\n[Open Folder](command:workbench.action.files.openFolder)\n You can also clone a repository from a URL. To learn more about how to use git and source control in VS Code [read our docs](https://aka.ms/vscode-scm).\n[Clone Repository](command:git.clone)',
```
Sorry..
But what should be the text there? any idea?

amiramw replied:
I'm not sure we have to register this view welcome for explorer. at some point we can just move to builtin git extension that configure it anyway, no?

EstherPerelman replied:
The second section will be removed and contribute by vsCode builtin git extension.

amiramw commented:
> > It works nicely, very good job. I added a few comments.
> > One question; if _Clone Repository_ is not supported when there is no opened workspace, why do we show it?
> > ![Screen Shot 2020-11-20 at 10 08 08](https://user-images.githubusercontent.com/1405703/99781907-c0df4800-2b18-11eb-8fbe-d6f59be66ab1.jpg)
> 
> @amiramw What do you think?

It should be supported when there is no opened workspace. At least in vscode it is with vscode builtin git support. Did you test with builtin git extension?

EstherPerelman commented:
> It works nicely, very good job. I added a few comments.
> 
> One question; if _Clone Repository_ is not supported when there is no opened workspace, why do we show it?
> 
> ![Screen Shot 2020-11-20 at 10 08 08](https://user-images.githubusercontent.com/1405703/99781907-c0df4800-2b18-11eb-8fbe-d6f59be66ab1.jpg)

@kittaakos Thanks for reviewing! I updated the code accordingly

kittaakos commented:
> It should be supported when there is no opened workspace

There was no workspace opened:
![Screen Shot 2020-11-23 at 13 34 41](https://user-images.githubusercontent.com/1405703/99962649-da8ac480-2d90-11eb-9b0f-1f7970b8de95.jpg)


> Did you test with builtin git extension?

I tested with the electron example app.

EstherPerelman commented:
> > It should be supported when there is no opened workspace
> 
> There was no workspace opened:
> ![Screen Shot 2020-11-23 at 13 34 41](https://user-images.githubusercontent.com/1405703/99962649-da8ac480-2d90-11eb-9b0f-1f7970b8de95.jpg)
> 
> > Did you test with builtin git extension?
> 
> I tested with the electron example app.

Anyway, I removed it and when working with the builtin git extension - it will be contributed by it (and supported).

amiramw commented:
@kittaakos are there any more comments or can we merge?

kittaakos commented:
> are there any more comments or can we merge?

I haven't had time to look into the updates, probably it is good to merge if all remarks have been addressed, but we need to restore the CI first.

amiramw commented:
@kittaakos PR is rebased and build passed. could you give final approval?

kittaakos reviewed (approved):
Thank you, @EstherPerelman. Nice feature👍 

@amiramw, please help with the merging. I've checked and tried the changeset twice already.

amiramw reviewed (approved):
Went over the code again and tested successfully
